ykrt/pass_finder/try_passes.py

```diff
@@ -185,18 +211,114 @@ def binary_split(logf, passes, is_prelink):
         print(72 * "=")
         print(list_of_passes_to_str(ok_passes))
 
-def main(logf, is_prelink):
+def evaluate_fitness(glogf, is_prelink, entity, passes, cwd):
+    try_passes = []
+    for i, bit in enumerate(entity):
+        if bit == 1:
+            try_passes.append(passes[i])
+    log(glogf, f"\ncurrently evaluating {try_passes}\n")
+    config = get_pipeline_config(is_prelink, try_passes)
+    ret, exec_time = test_pipeline(glogf, config, cwd) 
+    if ret:
+            exec_time = float(exec_time)  # Convert exec_time to a float
+            return  exec_time
+    else:
+        # Return False for programs that crash, and execution time as a large positive value
+        return float('inf')
+
+def crossover(parent1, parent2): 
+    crossover_point = random.randint(1, len(parent1) - 1)
+    child1 = parent1[:crossover_point] + parent2[crossover_point:]
+    child2 = parent2[:crossover_point] + parent1[crossover_point:]
+    return child1, child2
+
+def mutate(entity, mutation_rate):
+    mutated_entity = []
+    for bit in entity:
+        if random.random() < mutation_rate:
+            mutated_entity.append(1 - bit)  # Flip the bit
+        else:
+            mutated_entity.append(bit)
+    return mutated_entity
+
+def genetic_algorithm(glogf, is_prelink, population_size, mutation_rate, generations, target_fitness, passes, cwd):
+    population = []
+    fitness_scores = []
+
+    for _ in range(population_size):
+        entity = [random.randint(0,1) for _ in range(len(passes))]
+        population.append(entity)
+
+    for generation in range(generations):
+        log(glogf, "=========================================================")
+        log(glogf, f"\ngeneration: {generation}\n") 
+        fitness_scores.clear()
+        # Evaluate fitness for each entity in the population
+        fitness_scores = [evaluate_fitness(glogf, is_prelink, entity, passes, cwd) for entity in population]
+        log(glogf, f"\nfitness score: {fitness_scores}\n")
+        log(glogf, "=========================================================")
+
+        # less the execution time is better
```

Comment:
This sentence doesn't quite parse. How about "A lower execution time is better"?

Reply:
addressed here: 8dab2e70489ece37fe6efa8ab40983248dc19b1d